torch/csrc/jit/tensorexpr/llvm_codegen.cpp

```diff
@@ -229,6 +270,37 @@ class LLVMCodeGenImpl : public IRVisitor {
 
   void optimize(llvm::Module& M);
 };
+
+typedef void (*ParallelCallee)(int index, int8_t* packed_data);
+void DispatchParallel(int8_t* func, int start, int stop, int8_t* packed_data) {
+  // TODO: preserve the func type.
+  ParallelCallee callee = reinterpret_cast<ParallelCallee>(func);
+  // TODO: use absl::BlockingCounter or similar if less overhead is needed.
+  std::mutex m;
+  int remaining_tasks = stop - start; // GUARDED_BY m
+  std::condition_variable cv; // GUARDED_BY m
+  for (int index = start; index < stop; index++) {
+    if (index != stop - 1) {
+      at::intraop_launch([&, index]() {
+        callee(index, packed_data);
+        std::unique_lock<std::mutex> l(m);
+        remaining_tasks--;
+        cv.notify_one();
+      });
+      USE_TRIGGER(llvm_codegen_parallel_dispatched);
+    } else {
+      // the current thread handles the last one.
+      callee(index, packed_data);
+      USE_TRIGGER(llvm_codegen_parallel_dispatched);
+      std::unique_lock<std::mutex> l(m);
+      remaining_tasks--;
+      while (remaining_tasks != 0) {
+        cv.wait(l);
+      }
```

Comment:
I usually like to use the predicate version of `wait` e.g. `cv.wait(l, [&] { remaining == 0 })` but this is fine too